cpp/src/arrow/compute/kernels/scalar_string_test.cc

```diff
@@ -416,6 +424,28 @@ TYPED_TEST(TestStringKernels, SplitWhitespaceUTF8Reverse) {
                    &options_max);
 }
 
+#ifdef ARROW_WITH_RE2
+TYPED_TEST(TestStringKernels, ReplaceSubstringNormal) {
+  ReplaceSubstringOptions options{"foo", "bazz"};
+  this->CheckUnary("replace_substring", R"(["foo", "this foo that foo", null])",
+                   this->type(), R"(["bazz", "this bazz that bazz", null])", &options);
+  ReplaceSubstringOptions options_regex{"(fo+)\\s*", "\\1-bazz", -1};
+  this->CheckUnary("replace_substring_re2", R"(["foo ", "this foo   that foo", null])",
+                   this->type(), R"(["foo-bazz", "this foo-bazzthat foo-bazz", null])",
+                   &options_regex);
```

Comment:
Can you add a test with potential tricky cases? For example `text="aaaaaa", match="(a.a)", replacement="ab\1"`.